### ISSUE TITLE
Add a new profile that enables Link-Time Optimization (LTO)

### ISSUE THREAD

**Issue by gz** [feldera/feldera#2613]
From @zamazan4ik:

The `Cargo.toml` file Link-Time Optimization (LTO) for the Rust part of the project is not enabled. I suggest switching it on since it will reduce the binary size (always a good thing to have) and will likely improve the project's performance.

I suggest enabling LTO only for the Release builds so as not to sacrifice the developers' experience while working on the project since LTO consumes an additional amount of time to finish the compilation routine. If you think that a regular Release build should not be affected by such a change as well, then I suggest adding an additional `release-lto` profile where additionally to regular `release` optimizations LTO also will be added. Such a change simplifies life for maintainers and others interested in the project persons who want to build the most performant version of the application. Using ThinLTO also should help).

Additionally to LTO, I can suggest evaluating more advanced optimizations like Profile-Guided Optimization (PGO) and Post-Link Optimization (PLO). I have a lot of information (including a lot of benchmarks for many projects, including many databases!) in my repository: https://github.com/zamazan4ik/awesome-pgo . I expect measurable improvements from enabling PGO for Feldera. You can start to play with PGO with https://github.com/Kobzol/cargo-pgo

Thank you.

_Originally posted by @zamazan4ik in https://github.com/feldera/feldera/discussions/2609_

**Comment by ryzhyk** [comment]
Thin LTO is enabled by default in the release profile in Rust. Fat LTO makes compilation too slow for our use case without adding much performance. I have experimented with various other optimization settings, but it's hard to improve over Rust's default release config significantly.

**Comment by lalithsuresh** [comment]
Yeah given compilation times, I think thin LTO / release mode is good enough for now.

**Comment by zamazan4ik** [comment]
> Thin LTO is enabled by default in the release profile in Rust.

It's only partially true. The default LTO mode for the Release profile in Rust is `lto = false`. According to the [documentation](https://doc.rust-lang.org/cargo/reference/profiles.html#lto), it's a "thin local LTO", not a "thin LTO". The difference here is that "Thin LTO" performs LTO across different crates, "Thin local LTO" does not. It can be easily proven by enabling `lto = "thin"` for the Release profile and compiling a binary - you will get a different binary from the defaults.

> Fat LTO makes compilation too slow for our use case without adding much performance.

How much this "much" is? With Fat LTO a common way to avoid build overhead (that usually is a problem for local development and CI) is creating a dedicated profile like "heavy release mode", and build it only during actual releases (any other strategies are also fine).

> Yeah given compilation times, I think thin LTO / release mode is good enough for now.

However, if you think it's not worth it and you are okay with the current LTO - it's fine to leave the things as is.

By the way, what do you think about PGO? You only mentioned LTO above and not a word about PGO. I am wondering since you closed the issue. Have you already performed PGO tests for Feldera? If so, could you please share the results? If no, I think the issue should be reopened.

**Comment by gz** [comment]
Thanks, looking into PGO seems like a good idea.
